### PR TITLE
 [PAYMENT]: add service, dto, controller, repository

### PR DESCRIPTION
**Pull request by ConfusedSatlan**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -122,6 +122,12 @@
             <artifactId>junit-jupiter</artifactId>
             <scope>test</scope>
         </dependency>
+        <!-- https://mvnrepository.com/artifact/com.stripe/stripe-java -->
```
Comment (Denis-Balako) [inline pom.xml:125]: delete comment

File: src/main/java/com/team01/carsharingapp/controller/PaymentController.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/payment")
+public class PaymentController {
+    private final PaymentService paymentService;
+
+    @GetMapping
+    public ResponseEntity<List<PaymentDto>> getPayments(@RequestParam("user_id") Long userId) {
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/controller/PaymentController.java:23]: We don't use ResponseEntity in our project. Please, accord to project rule conventions.

File: src/main/java/com/team01/carsharingapp/controller/PaymentController.java
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/controller/PaymentController.java:1]: Check the link https://github.com/mate-academy/jv-car-sharing-service/tree/main#payments-endpoint
and implement features according to task.

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentDto.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.team01.carsharingapp.dto.payment;
+
+import com.team01.carsharingapp.model.Payment;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Data
+public class PaymentDto {
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/dto/payment/PaymentDto.java:8]: use records

File: src/main/java/com/team01/carsharingapp/mapper/PaymentMapper.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+@Mapper(config = MapperConfig.class)
+public interface PaymentMapper {
+    PaymentDto toDto(Payment payment);
+
+    Payment toEntity(PaymentRequestDto paymentRequestDto);
+
+    @AfterMapping
+    default void setOrderId(@MappingTarget PaymentDto paymentDto, Payment payment) {
+        paymentDto.setRentalId(payment.getRental().getId());
+    }
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/mapper/PaymentMapper.java:20]: Usage @ Mapping(target="...", source="...") for easy set more preferable

File: src/main/java/com/team01/carsharingapp/service/PaymentService.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.team01.carsharingapp.service;
+
+import com.team01.carsharingapp.dto.payment.PaymentDto;
+import com.team01.carsharingapp.dto.payment.PaymentRequestDto;
+import java.util.List;
+
+public interface PaymentService {
+    PaymentDto createPayment(PaymentRequestDto paymentRequestDto);
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/PaymentService.java:8]:
```suggestion
    PaymentDto createPayment(PaymentRequestDto requestDto);
```

Comment (Denis-Balako) [reply src/main/java/com/team01/carsharingapp/service/PaymentService.java:8]: rename in all classes

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+                .build();
+        try {
+            PaymentIntent paymentIntent = PaymentIntent.create(params);
+            String paymentIntentId = paymentIntent.getId();
+            Payment payment = new Payment();
+            BigDecimal amount = new BigDecimal(params.getAmount());
+            BigDecimal convertCent = new BigDecimal(CONVERT_CENT);
+            payment.setPrice(amount.divide(convertCent, 2, RoundingMode.HALF_UP));
+            payment.setType(Payment.Type.valueOf(paymentRequestDto.type()));
+            //Not finished
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:46]: Create //TODO comment or make a mock implementation

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        Stripe.apiKey = stripeSecretKey;
+        PaymentIntentCreateParams params = PaymentIntentCreateParams.builder()
+                .setAmount((paymentRequestDto.price()
+                        .multiply(new BigDecimal(CONVERT_CENT)).longValue()))
+                .setCurrency(paymentRequestDto.currency())
+                .setDescription("Car rental payment")
+                .addPaymentMethodType("cart")
+                .putMetadata("type", paymentRequestDto.type())
+                .build();
+        try {
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:38]: Just we need a huge try block here, or we can use it only for 1 line?

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            String paymentIntentId = paymentIntent.getId();
+            Payment payment = new Payment();
+            BigDecimal amount = new BigDecimal(params.getAmount());
+            BigDecimal convertCent = new BigDecimal(CONVERT_CENT);
+            payment.setPrice(amount.divide(convertCent, 2, RoundingMode.HALF_UP));
+            payment.setType(Payment.Type.valueOf(paymentRequestDto.type()));
+            //Not finished
+
+        } catch (StripeException e) {
+            e.getMessage();
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:49]: Create a custom exception for this case

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @Override
+    public PaymentDto getPayments() {
+        return null;
+    }
+
+    @Override
+    public List<PaymentDto> getPaymentsByUserId(Long userId) {
+        return null;
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:61]: NEVER return null. Just empty PaymentDto or empty List

**Review by Denis-Balako — CHANGES_REQUESTED** [review]
Good, but need some improvements

File: src/main/java/com/team01/carsharingapp/mapper/PaymentMapper.java (hunk 4 above)
Comment (ConfusedSatlan) [reply src/main/java/com/team01/carsharingapp/mapper/PaymentMapper.java:20]: it doesn't work like that with a value like that. I don’t know how I can get the ID from Rental and set to dto, is there either an additional function needed or what I wrote earlier, like we did in the book store

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentDto.java (hunk 3 above)
Comment (ConfusedSatlan) [reply src/main/java/com/team01/carsharingapp/dto/payment/PaymentDto.java:8]: i need it to set rentalId

File: src/main/java/com/team01/carsharingapp/controller/PaymentController.java
Comment (ConfusedSatlan) [reply src/main/java/com/team01/carsharingapp/controller/PaymentController.java:1]: I don't understand what u mean, for controller we have such as endpoints Payments Controller (Stripe): Facilitates payments for car rentals through the platform. Interacts with Stripe API. Use stripe-java library.

GET: /payments/?user_id=... - get payments
POST: /payments/ - create payment session
GET: /payments/success/ - check successful Stripe payments (Endpoint for stripe redirection)
GET: /payments/cancel/ - return payment paused message (Endpoint for stripe redirection) p.s. more information look at https://github.com/mate-academy/jv-car-sharing-service/tree/main#controllers

File: src/main/java/com/team01/carsharingapp/controller/PaymentController.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping(value = "/payment")
+public class PaymentController {
+    private static final String PAYMENT_SUCCESS = "Payment successful!";
+    private static final String PAYMENT_CANSEL = "Payment paused!";
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/controller/PaymentController.java:20]: Maybe cancelled?

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java [hunk 11]
```diff
@@ -0,0 +1,13 @@
+package com.team01.carsharingapp.dto.payment;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public record PaymentRequestDto(@NotBlank
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java:6]: Let's check for negative ids here
Also, add message for exception
```suggestion
public record PaymentRequestDto(@NotBlank(message = "id can't be null")
@Min(value = 0, message = "id can't be less than 0")
```

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+                                Long rentalId,
+                                @NotBlank
+                                @Size(min = 2, max = 3)
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java:9]: Add messages here

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 13]
```diff
@@ -1,63 +1,62 @@
...
+        currentRental.setReturnDate(LocalDate.now().plusDays(3));
+        //Values for stripe
+        BigDecimal totalPrice = calculateRentalCost(
+                currentRental.getRentalDate(),
+                currentRental.getReturnDate(),
+                currentRental.getCar().getDailyFee());
+        String currentCurrency = requestDto.currency();
+        String currentType = requestDto.type();
+        return new PaymentDto();
     }
```
Comment (Denis-Balako) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:42]: add empty lines for better readability

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+                                String currency,
+                                @NotBlank
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java:11]: Message here + @Size

**Review by Denis-Balako — APPROVED** [review]
Great!

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+public class PaymentServiceImpl implements PaymentService {
+    private final PaymentMapper paymentMapper;
+    private final Object rentalService;
+
+    @Override
+    public PaymentDto createPayment(PaymentRequestDto requestDto) {
+        //create test entity rental
+        Rental currentRental = new Rental();
+        currentRental.setId(requestDto.rentalId());
+        Car currentCar = new Car();
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:29]: You should create car here? Not get from DB?

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+        currentCar.setDailyFee(BigDecimal.valueOf(20.0));
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:30]: Magic number?

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+        currentRental.setCar(currentCar);
+        currentRental.setRentalDate(LocalDate.now());
+        currentRental.setReturnDate(LocalDate.now().plusDays(3));
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:33]: Magic number

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+        //Values for stripe
+        BigDecimal totalPrice = calculateRentalCost(
+                currentRental.getRentalDate(),
+                currentRental.getReturnDate(),
+                currentRental.getCar().getDailyFee());
+        String currentCurrency = requestDto.currency();
+        String currentType = requestDto.type();
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:40]: Why do you need it if you are not using it?

**Review by Someboty — CHANGES_REQUESTED** [review]
GJ! But I still don't get some parts...

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java (hunk 18 above)
Comment (ConfusedSatlan) [reply src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:40]: it's all variables for future impl stripe payment, i write comment in code

File: src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java (hunk 15 above)
Comment (ConfusedSatlan) [reply src/main/java/com/team01/carsharingapp/service/impl/PaymentServiceImpl.java:29]: i need this variables to imulate work class, when carService and rentalService will be impled, i change it, but now i must impl business logic

File: src/main/java/com/team01/carsharingapp/controller/PaymentController.java (hunk 10 above)
Comment (Someboty) [reply src/main/java/com/team01/carsharingapp/controller/PaymentController.java:20]: Still not fixed

File: src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java [hunk 19]
```diff
@@ -0,0 +1,17 @@
+package com.team01.carsharingapp.dto.payment;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.Size;
+
+public record PaymentRequestDto(@NotBlank(message = "id can't be null")
```
Comment (Someboty) [inline src/main/java/com/team01/carsharingapp/dto/payment/PaymentRequestDto.java:7]: NotBlank doesn't work with Long, use NotNull instead

**Review by Someboty — APPROVED** [review]
GJ! Don't forget to improve the propositions left.

**Pull request merged**